### ISSUE TITLE
image_dataset_from_directory supporting reshuffle_each_iteration

### ISSUE THREAD

**Issue by jamesbraza** [keras-team/tf-keras#354]
**System information**.

TensorFlow version (you are using): 20.10.0
Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state**.

I find [`tf.keras.utils.image_dataset_from_directory`](https://www.tensorflow.org/api_docs/python/tf/keras/utils/image_dataset_from_directory) very convenient for reading in images from data sources like Kaggle.  I use `shuffle=True` so that training subset sees a distribution of all classes.

The [`Dataset.shuffle`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shuffle) API supports an arg `reshuffle_each_iteration` that makes each `Dataset` iteration have the same shuffled ordering.  This `reshuffle_each_iteration = False` behavior currently is not supported by `image_dataset_from_directory`.

I am trying to compare the training process of several `Model`s.  Having the same ordering of shuffled images is valuable to me, for comparability.  Within one Python script, I would like to:
1. Read in a dataset using `image_dataset_from_directory`
2. Pass the same ordering of shuffled examples to 2+ `model.fit` calls

If `image_dataset_from_directory` supported `reshuffle_each_iteration = False`, I could enable this behavior within the call to `image_dataset_from_directory`.  However, since it's not currently supported, I have to manually reconstruct the behavior:
1. Read in the full dataset using `image_dataset_from_directory(..., shuffle=False)`
2. Calculate the size of the dataset: `dataset_size = len(list(dataset))`
    - Somewhat inefficient
3. Call `Dataset.shuffle(..., buffer_size=dataset_size, reshuffle_each_iteration=False)`

I think it would be useful if `image_dataset_from_directory` directly supported this behavior.

**Will this change the current api? How?**

It will add compatibility with `Dataset.shuffle(..., reshuffle_each_iteration=False)` to `image_dataset_from_directory`.

**Who will benefit from this feature?**

Users of `image_dataset_from_directory` who are looking to have repeatability across `Dataset` iterations.

**Comment by ThongLai** [comment]
+1 for adding this feature. I've encountered similar issues and want to share additional context that demonstrates why this feature is important:

Current Limitation with `validation_split`:
When using `image_dataset_from_directory` with `validation_split`, there's a tricky trade-off:
- With `shuffle=False`: Classes aren't evenly distributed between train/validation sets
- With `shuffle=True`: Classes are evenly distributed, but we lose control over the shuffling behavior for training set iterations

I tried implementing a manual shuffle like this:

```python
file_paths = tf.data.Dataset.from_tensor_slices(dataset.file_paths)
file_paths = file_paths.shuffle(buffer_size=file_paths.cardinality(), seed=42, reshuffle_each_iteration=False)
file_paths = [path.decode('utf-8') for path in file_paths.as_numpy_iterator()]

dataset = dataset.unbatch()
dataset = dataset.shuffle(buffer_size=len(file_paths), seed=42, reshuffle_each_iteration=False)
dataset = dataset.batch(32)
```

However, this approach requires setting `shuffle=False` in `image_dataset_from_directory`, which breaks the even class distribution in train/validation splits as said above.

Adding `reshuffle_each_iteration` support would solve these issues and provide better control over dataset shuffling while maintaining proper class distribution in splits.

**Note**: Regarding the dataset size calculation mentioned in the original post, there's a more efficient way:
Instead of `dataset_size = len(list(dataset))`, you can use `dataset_size = len(dataset.file_paths)` since `file_paths` is already a list attribute with the correct length as the same as the number of images in the dataset.
So I believe this solves the inefficient in the original post :). but my problem (using `validation_split` ) is still unsolvable :(
